### PR TITLE
[HOLD facebook/react-native#37465] Fix TextInput vertical alignment issue when using lineHeight prop on iOS (Paper - old arch)

### PR DESCRIPTION
**Pull request by fabOnReact**
# Upstream PR Link

Upstream PR https://github.com/facebook/react-native/pull/37465
fixes https://github.com/Expensify/App/issues/17767

## Summary:

Adding paragraphStyle.maximumLineHeight to a iOS UITextField displays the text under the UITextField ([ios-screenshot-1][1], [ios-screenshot-2][2], [ios-screenshot-3][3]). The issue reproduces on Storyboard App (iOS) using UITextField and  paragraphStyle.maximumLineHeight. It is not caused by react-native.

[1]: https://user-images.githubusercontent.com/24992535/238834159-566f7eef-ea2d-4fd4-a519-099b0a12046c.png "ios-screenshot-1"
[2]: https://user-images.githubusercontent.com/24992535/238834184-feb454a9-6504-4832-aec8-989f1d027861.png "ios-screenshot-2"
[3]: https://user-images.githubusercontent.com/24992535/238834283-cf572f94-a641-4790-92bf-bbe43afb1443.png "ios-screenshot-3"

The issue is caused by a private class _UITextLayoutFragmentView (a CALayer children of UITextField), which assumes the wrong position when setting the lineHeight. _UITextLayoutFragmentView frame's y coordinates are set to 30, instead of 0 ([react-native-screenshot-1][4], [react-native-screenshot-2][5])
- The _UITextLayoutFragmentView layer does not correctly position 
- The issue is caused by adding paragraphStyle.maximumLineHeight to UITextField.attributedText
- The parent UITextField bounds do not correctly position child _UITextLayoutFragmentView. 
The issue causes the below text Sdfsd to display under the TextInput.

[4]: https://github.com/Expensify/App/assets/24992535/06726b45-7e35-4003-9fcc-50c8d0dff0f6
[5]: https://github.com/Expensify/App/assets/24992535/d9745d29-8863-4170-bcc3-e78fa7e550d2

I was able to fix the issue and correctly align the private layout view _UITextLayoutFragmentView using the public API RCTUITextField textRectForBound, which allows modifying the UITextField frame and inset.
The solution consists in the following steps, applied only to UITextField with lineHeight prop:

1) set _UITextLayoutFragmentView to vertically align to the top. Required to correctly align _UITextLayoutFragmentView.
2) apply custom bounds with RCTUITextField textRectForBound to align _UITextLayoutFragmentView with the correct y coordinates and height
2) Adjust text baseline to be center aligned

fixes https://github.com/facebook/react-native/issues/28012

## Changelog:

[IOS] [FIXED] - Fix TextInput vertical alignment issue when using lineHeight prop on iOS (Paper - old arch)

## Test Plan:

Extensive tests included in the PR comments https://github.com/facebook/react-native/pull/37465#issuecomment-1551459879 and below

### CODE REVIEW

**Comment by fabOnReact** [comment]
<details><summary>After the fix - Testing RNTester in Expensify/react-native</summary>
<p>

https://github.com/Expensify/react-native/assets/24992535/30e9c55a-997d-495c-9db4-8337589137fc




</p>
</details>

**Comment by fabOnReact** [comment]
1st Test - Verify that TextInput vertical alignment is fixed and text is not cut

| Before    | After |
| ----------- | ----------- |
| <image src="https://github.com/Expensify/react-native/assets/24992535/b6a871d7-679f-4f96-8d27-f8955a550052" width="1000" />      | <image src="https://github.com/Expensify/react-native/assets/24992535/5f72b882-2bc6-43fe-a477-321c5312b0f1" width="1000" />       |

2nd Test - No regression detected

| Before    | After |
| ----------- | ----------- |
| <video src="https://github.com/Expensify/react-native/assets/24992535/0e08769f-ecc7-4d2f-804e-e6671ced847c" width="350" />      | <video src="https://github.com/Expensify/react-native/assets/24992535/d46b9943-5b1b-4e4f-90d8-482c2a9720b7" width="350" />       |

File: packages/react-native/Libraries/Text/RCTTextAttributes.m
Comment (cead22) [inline packages/react-native/Libraries/Text/RCTTextAttributes.m]: Can you capitalize all comments and add a line break before them (except for the one after the if which is in a new level of indentation)?
```suggestion
    // Text with lineHeight lower then font.lineHeight does not correctly vertically align
```

**Review by cead22 — COMMENTED** [review]
Code looks good, but I'd love to get a review from someone that's more familiar with this code and how it works

File: packages/react-native/Libraries/Text/RCTTextAttributes.m
Comment (fabOnReact) [reply packages/react-native/Libraries/Text/RCTTextAttributes.m]: Thanks @cead22. I made the changes with commit https://github.com/Expensify/react-native/pull/57/commits/12b51ddb8b995e4a08278d5094c6d2c838c09f04. I remain available for further improvements.

File: packages/react-native/Libraries/Text/TextInput/Singleline/RCTUITextField.m
Comment (fedirjh) [inline packages/react-native/Libraries/Text/TextInput/Singleline/RCTUITextField.m]:
```suggestion
    // Apply custom bounds to fix iOS UITextField issue with lineHeight.
    // Sets the correct y coordinates for _UITextLayoutFragmentView.
    return UIEdgeInsetsInsetRect([super textRectForBounds:self.fragmentViewContainerBounds.size.height > 0 ? self.fragmentViewContainerBounds : bounds], borderAndPaddingInsets);
```

Can’t be simplified ?

Comment (fabOnReact) [reply packages/react-native/Libraries/Text/TextInput/Singleline/RCTUITextField.m]: Thanks a lot, @fedirjh. I applied the changes with commits https://github.com/Expensify/react-native/pull/57/commits/613bf0a29dc3c68525fa8f1c193aa2e4e7350720, https://github.com/Expensify/react-native/pull/57/commits/a5ac3c31c59e22c99991b4fb132d6a87cb3d56bc and tested the functionality again


<details><summary>CLICK TO OPEN TESTS RESULTS</summary>
<p>

<image src="https://github.com/Expensify/react-native/assets/24992535/9ba4dc0f-6fcd-41b4-817b-3fad7d3d6487" width="250" />


</p>
</details>

<details><summary>CLICK TO OPEN TESTS RESULTS</summary>
<p>

<image src="https://github.com/Expensify/react-native/assets/24992535/4f1b3859-1de9-4c1e-8d3c-ba545a124082" width="250" />

</p>
</details>

**Review by j-piasecki — DISMISSED** [review]
LGTM

File: packages/react-native/Libraries/Text/TextInput/Singleline/RCTUITextField.m
Comment (cead22) [inline packages/react-native/Libraries/Text/TextInput/Singleline/RCTUITextField.m]: Minor, so not a blocker
```suggestion

  UIEdgeInsets borderAndPaddingInsets = UIEdgeInsetsMake(_textContainerInset.top, leftPadding, _textContainerInset.bottom, rightPadding);
  // The fragmentViewContainerBounds set the correct y coordinates for
```

**Review by cead22 — DISMISSED** [review]
Looks good to me. @fedirjh can you re-review when you get a chance please?

**Review by fedirjh — DISMISSED** [review]
LGTM and tests well 

@cead22 which branch should we target ?

File: packages/react-native/Libraries/Text/TextInput/Singleline/RCTUITextField.m
Comment (fabOnReact) [reply packages/react-native/Libraries/Text/TextInput/Singleline/RCTUITextField.m]: Thanks @cead22. 

#### Adding a line-break before the comment
I believe you refer to
>Can you capitalize all comments and add a line break before them (except for the one after the if which is in a new level of indentation)?

I added a line break before the comment with commit https://github.com/Expensify/react-native/pull/57/commits/3d12d11080170c91407676576eac59b90afaf082.

#### Github Website Issue

There was an issue with the GitHub (website), which would not update the PR Diff and commits. My remote branch [fabriziobertoglio1987/text-input-vertical-alignment-expensify-2](https://github.com/fabriziobertoglio1987/react-native/tree/text-input-vertical-alignment-expensify-2) included changes not available on Github Website PR.


<details><summary>CLICK TO OPEN VIDEO</summary>
<p>

https://github.com/Expensify/react-native/assets/24992535/7e5750ca-eaa8-4f14-b2c6-2294fca2f450

</p>
</details>

For this reason, I had to temporarily change the base branch.

Related Discussions: https://github.com/orgs/community/discussions/16351 https://github.com/isaacs/github/issues/750#issuecomment-410536223

I tested and reviewed the branch diff locally (`git checkout text-input-vertical-alignment-expensify-2 && g diff ExpensifyRC1-0.72.0-alpha.0..`), and the branch has no issues.

<details><summary>CLICK TO OPEN TESTS RESULTS</summary>
<p>


Screenshot 2023-06-01 at 2 47 34 PM

</p>
</details>

Sorry. I remain available. Thanks

**Comment by cead22** [comment]
Out of curiosity, why not merge into main?

**Comment by fedirjh** [comment]
It looks like `ExpensifyRC1-0.72.0-alpha.0` will be released. We have other PRs are targeting the same branch,  check this comment https://github.com/Expensify/react-native/pull/58#issuecomment-1569139541

**Comment by fabOnReact** [comment]
The main branch facebook/react-native changes every day. The facebook/react-native main branch sometimes has issues, for this reason, it is easier to use the release branch.
The release branches are tested from the community (for example see https://github.com/reactwg/react-native-releases/discussions/54 for 0.72).
The Expensify team is working to upgrade react-native to RC1-0.72.0-alpha.0 see discussion https://github.com/Expensify/App/pull/18507

**Review by cead22 — APPROVED** [review]

**Comment by cead22** [comment]
Holding until the FB PR is approved and merged

**Comment by cead22** [comment]
@fabriziobertoglio1987 any idea if/when the upstream PR will be merged?

**Comment by fabOnReact** [comment]
Thanks @cead22. The PR will be merged. I reached out via Discord to receive feedback on the PR.
I will work until the PR is merged. I will keep you updated about the progress.

**Comment by fabOnReact** [comment]
Thanks @cead22 

This is the last update from Samuel Susla https://github.com/facebook/react-native/pull/37465#issuecomment-1600505344
>Sorry for the slow reply, I was away for +2 weeks on vacation. I will just fix the linter issues, they are only related to code formatting.

The was an issue with the internal facebook code linter (not open source). The PR will be merged soon. I remain available.

**Comment by cead22** [comment]
Awesome, thanks for the update!

**Comment by fedirjh** [comment]
cc @cead22 @fabriziobertoglio1987 The upstream PR was merged today 🎉 🎉 
- https://github.com/facebook/react-native/pull/37465#issuecomment-1621700252

**Comment by cead22** [comment]
@fedirjh @j-piasecki can you re-review this please? I think we're ready to remove the hold, but to confirm, is this PR still targeting the right branch?

**Comment by fabOnReact** [comment]
This are the information I found:
- **Expensify/react-native**: Expensify is updating react-native to 0.72 https://github.com/Expensify/App/pull/18507#issuecomment-1623610848
- **Expensify/react-native**: Expensify is merging a list of selected PRs in their Expensify/react-native fork to be included in the 0.72 expensify/react-native release (see this [discussion](https://github.com/Expensify/App/pull/18507#issuecomment-1621573098)). We may ask to cherry pick this PR in that discussion.
- **facebook/react-native**: Cherry pick request for this [commit](https://github.com/facebook/react-native/commit/35a1648d0c93ef2db1de6b8dbe7fc1d997c24039) could be done in the next release 0.73 https://github.com/reactwg/react-native-releases/discussions/64. Cherry picks in [0.72.2](https://github.com/reactwg/react-native-releases/discussions/76) are for [regressions](https://reactnative.dev/contributing/release-faq#what-is-a-qualified-pick-request) introduced with 0.72.
- **Expensify/react-native**: React Native and Expensify did not yet prepare 0.73 branch. We can not target main, as Expensify/react-native is a fork of facebook/react-native.

**Comment by j-piasecki** [comment]
@cead22 It looks like all commits since the last review are cosmetic or done simply to trigger CI, so should be good 🚀

**Comment by fabOnReact** [comment]
#### Update 07/07

- Investigating regression reported in comment https://github.com/facebook/react-native/pull/37465#issuecomment-1625110372

**Comment by fabOnReact** [comment]
### Update 08/07

- Published DRAFT PR [#38258](https://github.com/facebook/react-native/pull/38258) to fix [regression](https://github.com/facebook/react-native/pull/37465#issuecomment-1625110372) reported by Samuel Susla.

**Comment by fabOnReact** [comment]
### Update 11/07
- Further reviewed the solutions from PRs [#38258](https://github.com/facebook/react-native/pull/38258) and [37465](https://github.com/facebook/react-native/pull/37465)
- Testing different edge cases
- Investigating the root cause of the [regression](https://github.com/facebook/react-native/pull/37465#issuecomment-1625110372) and evaluating an alternative implementation.

**Comment by fabOnReact** [comment]
### Update 12/07

There is a difference in the baseline computation between RCTTextShadowView [postprocessAttributedText](https://github.com/facebook/react-native/blob/9ab27e8895d6934e72ebdc601d169578ab9628f1/packages/react-native/Libraries/Text/Text/RCTTextShadowView.m#L124-L168) and RCTTextAttributes [effectiveParagraphStyle](https://github.com/facebook/react-native/pull/37465/files#diff-38333c03094e568d775af03727306a9e737fec7434e31f2f6e849f9663e48052R181-R184).

Given the following example:

```xml
<Text style={{lineHeight: 500, fontSize: 15, backgroundColor: 'red'}}>
  font size 15
  <Text style={{fontSize: 50, lineHeight: 500}}>font size 50</Text>
</Text>
```

- The RCTTextShadowView implementation calculates the same baseline for the entire Text based on the highest lineHeight/fontSize
- The RCTTextAttributes implementation calculates different baseline for each Text span, based on the Text span lineHeight and fontSize

**Comment by fabOnReact** [comment]
### Update 22/07
* Moved [Fix TextInput vertical alignment issue when using lineHeight prop on iOS without changing Text baseline (Paper - old arch)  facebook/react-native#38359](https://github.com/facebook/react-native/pull/38359) to ready to review

**Comment by cead22** [comment]
@fabOnReact any update on this one?

**Comment by fedirjh** [comment]
@cead22 I think we should close this PR. This fork is not used anymore in newDot.

**Comment by fabOnReact** [comment]
The upstream PR was merged. I'm closing this PR. Thanks.

**Pull request closed without merging**
